### PR TITLE
HWKAPM-809 Improve calculation of 'actual' time to cater for concurre…

### PR DESCRIPTION
**Pull request by objectiser**
…nt child nodes

### CODE REVIEW

**Comment by objectiser** [comment]
@jpkrohling Could you review please? Previously if the child nodes accumulated durations was deducted from the parent's duration, it could result in a negative 'actual' time if the child nodes were associated with concurrent activities. So the updated algorithm checks for this situation, and then attempts to work out the actual time based on obtaining the earliest start time and latest end time for the child nodes. If that is still longer than the parent's duration, then the nodes represent async activities outliving the parent, so the parent's actual time is set to the same as the elapsed.

File: server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java [hunk 1]
```diff
@@ -164,4 +158,44 @@ protected void deriveNodeDetails(Trace trace, List<Node> nodes, List<NodeDetails
...
+     * performed concurrently and therefore not impacting the time spent in the parent node
+     * then this method will return an elapsed time of 0.
+     *
+     * @param n The parent node
+     * @return The elapsed time of the child nodes
+     */
+    protected long calculateChildElapsedTime(Node n) {
+        long childElapsed = 0;
+        if (n.containerNode()) {
+            long startTime=n.getTimestamp() + n.getDuration();
```
Comment (jpkrohling) [inline server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java]: Is this the "minimum" start time expected for the next service? Or is it the actual start time for this node? For a "startTime", I'd expect it to be the timestamp , or timestamp minus duration, not plus.

File: server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java [hunk 2]
```diff
@@ -164,4 +158,44 @@ protected void deriveNodeDetails(Trace trace, List<Node> nodes, List<NodeDetails
...
+            long endTime = n.getTimestamp();
+            for (int j = 0; j < ((ContainerNode) n).getNodes().size(); j++) {
```
Comment (jpkrohling) [inline server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java]: Could you rename `j` to `i`? With `j` as the index, I keep looking for the outer `for`, which doesn't exist :)

File: server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java [hunk 3]
```diff
@@ -105,13 +105,7 @@ protected void deriveNodeDetails(Trace trace, List<Node> nodes, List<NodeDetails
                 nd.setCorrelationIds(n.getCorrelationIds());
                 nd.setElapsed(n.getDuration());
```
Comment (pavolloffay) [inline server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java]: From docs: 
`The elapsed time represents the total time spent within a particular component (including calls to child components)`
maybe better to change it to span finish - span start.

File: server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java (hunk 1 above)
Comment (objectiser) [inline server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java]: The timestamp is the start time of the node - so adding the duration identifies the end time of the node - but this is for working backwards to find the earliest start time of the contained child nodes.

**Comment by objectiser** [comment]
@jpkrohling yes you are correct, it would return 2. As this is based on the simple assumption that child activities are synchronous. The only other option would be to check if the child activities are overlapping, which may be more intensive, but more accurate?

File: server/processors/src/test/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriverTest.java [hunk 4]
```diff
@@ -170,4 +170,67 @@ public void testInitial() throws RetryAttemptException {
...
+        component1.setTimestamp(1100);
+        component1.setDuration(200);
+        consumer.getNodes().add(component1);
+
+        Component component2 = new Component();
+        component2.setTimestamp(1300);
+        component2.setDuration(200);
+        consumer.getNodes().add(component2);
+
+        assertEquals(400, deriver.calculateChildElapsedTime(consumer));
```
Comment (pavolloffay) [inline server/processors/src/test/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriverTest.java]: This is correct, however what if children are started in new thread (asynchronously) then node representing parent node can perform other operations, therefore actual time would be parent node `end-start time`

From docs `he actual time represents only the time specific in the component itself.` What is the benefit of `actual time`?

File: server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java [hunk 5]
```diff
@@ -164,4 +158,44 @@ protected void deriveNodeDetails(Trace trace, List<Node> nodes, List<NodeDetails
         commonProperties.addAll(trace.getProperties(Constants.PROP_PRINCIPAL));
         return commonProperties;
     }
+
+    /**
+     * This method calculates the elapsed time associated with the child nodes of the
+     * supplied parent node. If the supplied node is not a parent, or if the child nodes are
```
Comment (objectiser) [inline server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java]: I'll update the comment - basically just means that if the supplied node has no child nodes, then it elapsed time  will equal actual time.

File: server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java [hunk 6]
```diff
@@ -164,4 +158,44 @@ protected void deriveNodeDetails(Trace trace, List<Node> nodes, List<NodeDetails
...
+     * performed concurrently and therefore not impacting the time spent in the parent node
+     * then this method will return an elapsed time of 0.
+     *
+     * @param n The parent node
```
Comment (objectiser) [inline server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java]: I'll remove parent in the comment.

File: server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java (hunk 1 above)
Comment (objectiser) [inline server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java]: They are being processed independently, and unless there is something wrong with the data provided in the child nodes,then after processing the child nodes the start time will be less than the end time.

File: server/processors/src/test/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriverTest.java (hunk 4 above)
Comment (objectiser) [inline server/processors/src/test/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriverTest.java]: The benefit of the actual time is when the child nodes are performed in its scope, to differentiate where the time is spent. The problem is when to use a simple approach for calculating this value, versus understanding that the children are async and therefore a more complex approach may need to be used.

**Comment by pavolloffay** [comment]
Personally, this `acutal` time was always bit strange thing for me, as there are a couple of problems with its calculation (e.g. async descendants, for producer actual=elapsed, if there is no component after consumer actual=elapsed(not 100% sure here)). I'm not sure if users would rely on this value.

File: server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java (hunk 3 above)
Comment (pavolloffay) [inline server/processors/src/main/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriver.java]: yes something like that (if it makes sense :))

**Comment by objectiser** [comment]
@jpkrohling @pavolloffay Agree that a tooltip should be added to make clear that this value may not be 100% accurate and provide explanation of approach somewhere (possibly not in the tooltip as might be too detailed). However I believe it can be useful to indicate situations where not all of the elapsed time is spent in that component.

**Comment by objectiser** [comment]
@jpkrohling @pavolloffay I want to revisit the algorithm, but in the meantime would this updated text work? https://github.com/hawkular/hawkular-apm-user-guide/pull/28

**Comment by jpkrohling** [comment]
I thought you'd apply this change, and revisit the algorithm in a later revision. As per JIRA, it seems you want to apply it for this PR. So, I'm canceling my review.

**Comment by objectiser** [comment]
I think this change is ok - as it fixes the problem. So we could merge as is (possibly with the additional test) and then consider further changes later. @pavolloffay would that be ok with you?

File: server/processors/src/test/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriverTest.java [hunk 7]
```diff
@@ -170,4 +170,67 @@ public void testInitial() throws RetryAttemptException {
         assertFalse(details.get(1).isInitial());
     }
 
+    @Test
+    public void testCalculateActualTimeSync() throws RetryAttemptException {
```
Comment (pavolloffay) [inline server/processors/src/test/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriverTest.java]: Could you please add test, for which holds:

* one parent, one child 
* child duration < parent
* child ends after parent node.

Comment (objectiser) [inline server/processors/src/test/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriverTest.java]: `testCalculateActualTimeAsync` I believe satisfies the second and third cases? the second child duration (700) is greater than the parent's duration (500), and consequently it ends after the parent node (your third case).

The test `testCalculateActualTimeSync` has two children, one after the other - I don't mind adding a test for the first case, but wondering if there was a special reason?

Comment (pavolloffay) [inline server/processors/src/test/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriverTest.java]: > testCalculateActualTimeAsync I believe satisfies the second and third cases?

Yes, I'm thinking about this as white box testing. The second point is not satisfied:  in ` testCalculateActualTimeAsync` sum of elapsed times greater than parent's duration.

Comment (objectiser) [inline server/processors/src/test/java/org/hawkular/apm/server/processor/nodedetails/NodeDetailsDeriverTest.java]: Good catch - PR updated. If ok, then I'll rebase.

**Comment by objectiser** [comment]
retest this please

**Comment by objectiser** [comment]
@jpkrohling @pavolloffay Thanks!

**Pull request merged**